Sources/PriorityQueueModule/PriorityQueue.swift

```diff
@@ -0,0 +1,168 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift Collections open source project
+//
+// Copyright (c) 2021 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+//
+//===----------------------------------------------------------------------===//
+
+import Swift
+
+/// A double-ended priority queue built on top of `Heap`.
+public struct PriorityQueue<Value, Priority: Comparable> {
+  public typealias Pair = (value: Value, priority: Priority)
+
+  @usableFromInline
+  struct _Element: Comparable {
+    @usableFromInline let value: Value
+    let priority: Priority
+    let insertionCounter: UInt64
+
+    @usableFromInline
+    init(value: Value, priority: Priority, insertionCounter: UInt64) {
+      self.value = value
+      self.priority = priority
+      self.insertionCounter = insertionCounter
+    }
+
+    @usableFromInline
+    static func == (lhs: Self, rhs: Self) -> Bool {
+      lhs.priority == rhs.priority
+    }
+
+    @usableFromInline
+    static func < (lhs: Self, rhs: Self) -> Bool {
+      if lhs.priority < rhs.priority {
+        return true
+      } else if lhs.priority == rhs.priority {
```

Comment:
One issue we've run into in our upstream implementation of this, is that this `Comparable` implementation only works when trying to dequeue the minimum element. Dequeueing the maximum element doesn't respect FIFO ordering due to the default implementation of `>` being to flip the arguments and use `<` (i.e. `a > b` ⇢ `b < a`). Implementing `>` to also account for the insertion counter starts violating invariants in the heap — it creates weird situations where an element can be both greater than and less than another if it was inserted before and has the same priority.

One way we've come up with to fix this is to only insert the priorities into the underlying heap and use them as a key for looking up an array (or linked list or deque) in a dictionary that actually contains the values. That eliminates the need to store a `UInt64` per element but comes at some cost (especially if you use an array) and requires that the `Priority` be `Hashable`. None of these changes affect the underlying `Heap`.

```swift
struct PriorityQueue<Value, Priority: Hashable> {
    var _base: Heap<Priority>
    var _elements: Dictionary<Priority, Deque<Value>>

    ...
}
```

Are there other approaches for maintaining FIFO ordering?